src/celeritas/io/RootEventReader.cc

```diff
@@ -0,0 +1,101 @@
+#include "RootEventReader.hh"
+
+#include <TFile.h>
+#include <TLeaf.h>
+#include <TTree.h>
+
+#include "corecel/io/Logger.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with ROOT input filename.
+ */
+RootEventReader::RootEventReader(std::string const& filename,
+                                 SPConstParticles params)
+    : params_(std::move(params))
+{
+    CELER_EXPECT(!filename.empty());
+    tfile_.reset(TFile::Open(filename.c_str(), "read"));
+    CELER_ASSERT(tfile_->IsOpen());
+    ttree_.reset(tfile_->Get<TTree>(tree_name()));
+    CELER_ASSERT(ttree_);
+
+    num_entries_ = ttree_->GetEntries();
+    CELER_ASSERT(num_entries_ > 0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read single event from the primaries tree.
+ */
+auto RootEventReader::operator()() -> result_type
+{
+    CELER_EXPECT(entry_count_ <= num_entries_);
+    ttree_->GetEntry(entry_count_);
+    auto const this_evt_id = ttree_->GetLeaf("event_id")->GetValue();
+
+    result_type primaries;
+    for (; entry_count_ < num_entries_; entry_count_++)
+    {
+        ttree_->GetEntry(entry_count_);
+        if (ttree_->GetLeaf("event_id")->GetValue() != this_evt_id)
+        {
+            break;
+        }
+
+        Primary primary;
+        primary.event_id = EventId{this->from_leaf<size_type>("event_id")};
+        primary.track_id = TrackId{this->from_leaf<size_type>("track_id")};
+        primary.particle_id
+            = params_->find(PDGNumber{this->from_leaf<int>("particle")});
+        primary.energy
+            = units::MevEnergy{this->from_leaf<real_type>("energy")};
+        primary.time = this->from_leaf<real_type>("time");
+        primary.position = this->from_array_leaf("pos");
+        primary.direction = this->from_array_leaf("dir");
+        primaries.push_back(std::move(primary));
+    }
+
+    CELER_LOG_LOCAL(info) << "Read event " << this_evt_id << " with "
+                          << primaries.size() << " primaries";
+    return primaries;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to fetch leaves.
+ */
+template<class T>
+auto RootEventReader::from_leaf(char const* leaf_name) -> T
+{
+    CELER_EXPECT(ttree_);
+    auto const leaf = ttree_->GetLeaf(leaf_name);
+    CELER_ASSERT(leaf);
+    return static_cast<T>(leaf->GetValue());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to fetch leaves containing `std::array<double, 3>`.
+ */
+Real3 RootEventReader::from_array_leaf(char const* leaf_name)
+{
+    CELER_EXPECT(ttree_);
+    auto const leaf = ttree_->GetLeaf(leaf_name);
+    CELER_ASSERT(leaf);
+    CELER_ASSERT(leaf->GetLen() == 3);
+    return {leaf->GetValue(0), leaf->GetValue(1), leaf->GetValue(2)};
+}
```

Comment:
Depending on how often we write a non-dictionary ROOT file it might be h
```c++
class RootLeafReader
{
explicit RootLeafReader(TTree* tree);

template<class T>
void operator()(char const* leaf_name, T* value) { /* default implementation */ }

template<class T, class N>
void from_leaf(char const* leaf_name, Array<T, N>* value) { /* call GetValue on each item */ }

template<class T>
void from_leaf(char const* leaf_name, OpaqueId<T>* value) {...}
```

so that you can make a local `RootLeafReader read_leaf{ttree_.get()}; read_leaf("pos", p.pos);` etc.

Probably not worth it for this small file but something to keep in mind.